### ISSUE TITLE
Distributed training with wandb.sweep

### ISSUE THREAD

**Issue by LTayfaker** [wandb/wandb#7811]
**I'm using wandb for the first time and I'm having a bit of a problem, when running a model with a single multi-GPU, I want to use wandb.sweep to scan for hyperparameters, I'm guessing that the code is running with a new sweep created on top of each GPU due to the part of distributed training being run first and then initializing wandb.sweep, what I'm hoping for I want to have only one sweep_id and train my model on multiple GPUs, what should I do?**

**The environment I use is pytorch 2.0 and CUDA version 12.2,I trained my model by typing the command "torchrun --nproc_per_node=6 train.py" in the terminal.**

**Here is the main function entry for my model's train.py code:**
```
if __name__ == "__main__":
    sweep_config = get_wandb_config()
    sweep_id = wandb.sweep(sweep_config, project=sweep_config['project_name'])
    wandb.agent(sweep_id, run, count=3)
```

**Here is the main code for the run function：**
```
def run():
    .....
    args = get_args_parser()
    nowtime = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    wandb.init(project=sweep_config['project_name'], name=nowtime, save_code=True, config=sweep_config, group="DDP")
    args.learning_rate = wandb.config.learning_rate
    args.lr_drop_epoch = wandb.config.lr_drop_epoch
    args.max_epoch_num = wandb.config.max_epoch_num
    args.optim = wandb.config.optim
    net = MaskDecoderHQ(args.model_type)
    main(net, train_datasets, valid_datasets, args, sweep_config)
    wandb.finish()
```

**Here is the code for the first part of the train function：**
```
def main(net, train_datasets, valid_datasets, args, sweep_config):
    misc.init_distributed_mode(args)
    seed = args.seed + misc.get_rank()
    torch.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)
    ......
```

**Then there is the output log of the model run:**

```
wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways.
wandb: WARNING To avoid this, please fix the sweep config schema violations below:
wandb: WARNING   Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected)
Create sweep with ID: rqbqivtl
Sweep URL: https://wandb.ai/.../sweeps/rqbqivtl
wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways.
wandb: WARNING To avoid this, please fix the sweep config schema violations below:
wandb: WARNING   Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected)
Create sweep with ID: 31foef6f
Sweep URL: https://wandb.ai/.../sweeps/31foef6f
wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways.
wandb: WARNING To avoid this, please fix the sweep config schema violations below:
wandb: WARNING   Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected)
Create sweep with ID: 27tb9vkv
Sweep URL: https://wandb.ai/.../sweeps/27tb9vkv
wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways.
wandb: WARNING To avoid this, please fix the sweep config schema violations below:
wandb: WARNING   Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected)
Create sweep with ID: 27m6nd76
Sweep URL: https://wandb.ai/../sweeps/27m6nd76
wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways.
wandb: WARNING To avoid this, please fix the sweep config schema violations below:
wandb: WARNING   Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected)
Create sweep with ID: b0f81ait
Sweep URL: https://wandb.ai/.../sweeps/b0f81ait
wandb: Agent Starting Run: sofplkis with config:
wandb: 	learning_rate: 0.0005
wandb: 	lr_drop_epoch: 80
wandb: 	max_epoch_num: 200
wandb: 	optim: Adam
wandb: Agent Starting Run: o5mmxwkd with config:
wandb: 	learning_rate: 0.0003
wandb: 	lr_drop_epoch: 80
wandb: 	max_epoch_num: 100
wandb: 	optim: Adam
wandb: Agent Starting Run: 2y8wyoo9 with config:
wandb: 	learning_rate: 0.0001
wandb: 	lr_drop_epoch: 40
wandb: 	max_epoch_num: 150
wandb: 	optim: Adam
wandb: Agent Starting Run: mewgkh87 with config:
wandb: 	learning_rate: 0.0001
wandb: 	lr_drop_epoch: 100
wandb: 	max_epoch_num: 100
wandb: 	optim: SGD
wandb: Agent Starting Run: 0ukaa50c with config:
wandb: 	learning_rate: 0.0005
wandb: 	lr_drop_epoch: 100
wandb: 	max_epoch_num: 150
wandb: 	optim: SGD
wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways.
wandb: WARNING To avoid this, please fix the sweep config schema violations below:
wandb: WARNING   Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected)
Create sweep with ID: lqiseifr
Sweep URL: https://wandb.ai/.../sweeps/lqiseifr
wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: Agent Starting Run: mjo15l5x with config:
wandb: 	learning_rate: 0.0004
wandb: 	lr_drop_epoch: 60
wandb: 	max_epoch_num: 200
wandb: 	optim: SGD
wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: - Waiting for wandb.init()...
wandb: - Waiting for wandb.init()...
wandb: \ Waiting for wandb.init()...
wandb: \ Waiting for wandb.init()...
wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: Tracking run with wandb version 0.17.1
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run 2024-06-18 02:10:19
wandb: ⭐️ View project at https://wandb.ai/...
wandb: 🧹 View sweep at https://wandb.ai/.../sweeps/lqiseifr
wandb: 🚀 View run at https://wandb.ai/.../runs/mjo15l5x
wandb: Tracking run with wandb version 0.17.1
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run 2024-06-18 02:10:17
wandb: ⭐️ View project at https://wandb.ai/...
wandb: 🧹 View sweep at https://wandb.ai/.../sweeps/31foef6f
wandb: 🚀 View run at https://wandb.ai/.../runs/o5mmxwkd
```

**From the output logs, my guess is that each GPU running the model code creates a new sweat id, and each sweat id in turn scans all the hyperparameter combinations I've set up, but I actually only need one sweat id that scans a set of hyperparameter combinations, but since I'm going to be going into my run in the sweat.agent() function part of the code function part of the code.I think it's the order, but there's no solution yet!**

**Comment by LTayfaker** [comment]
> Jason Davenport commented: Hello! To make sure I can best assist you, may I please ask for more information? A copy of the code in question would be the best place to start, along with a detailed rundown of what environment you are running in.
> 
> Let me know! Thanks 🙂
Hi, I've updated my question, can you look at it for me or is there any more information I need to provide?

**Comment by LTayfaker** [comment]
> I'm using wandb for the first time and I'm having a bit of a problem, when running a model with a single multi-GPU, I want to use wandb.sweep to scan for hyperparameters, I'm guessing that the code is running with a new sweep created on top of each GPU due to the part of distributed training being run first and then initializing wandb.sweep, what I'm hoping for I want to have only one sweep_id and train my model on multiple GPUs, what should I do?
> 
> The environment I use is pytorch 2.0 and CUDA version 12.2,I trained my model by typing the command "torchrun --nproc_per_node=6 train.py" in the terminal.
> 
> Here is the main function entry for my model's train.py code: if **name** == "**main**": sweep_config = get_wandb_config() sweep_id = wandb.sweep(sweep_config, project=sweep_config['project_name']) wandb.agent(sweep_id, run, count=3)
> 
> ### Here is the main code for the run function： def run(): ..... args = get_args_parser() nowtime = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S") wandb.init(project=sweep_config['project_name'], name=nowtime, save_code=True, config=sweep_config, group="DDP") args.learning_rate = wandb.config.learning_rate args.lr_drop_epoch = wandb.config.lr_drop_epoch args.max_epoch_num = wandb.config.max_epoch_num args.optim = wandb.config.optim net = MaskDecoderHQ(args.model_type) main(net, train_datasets, valid_datasets, args, sweep_config) wandb.finish()
> 
> Here is the code for the first part of the train function： def main(net, train_datasets, valid_datasets, args, sweep_config): misc.init_distributed_mode(args) seed = args.seed + misc.get_rank() torch.manual_seed(seed) np.random.seed(seed) random.seed(seed) ......
> 
> Then there is the output log of the model run:
> 
> wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways. wandb: WARNING To avoid this, please fix the sweep config schema violations below: wandb: WARNING Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected) Create sweep with ID: rqbqivtl Sweep URL: https://wandb.ai/.../sweeps/rqbqivtl wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways. wandb: WARNING To avoid this, please fix the sweep config schema violations below: wandb: WARNING Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected) Create sweep with ID: 31foef6f Sweep URL: https://wandb.ai/.../sweeps/31foef6f wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways. wandb: WARNING To avoid this, please fix the sweep config schema violations below: wandb: WARNING Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected) Create sweep with ID: 27tb9vkv Sweep URL: https://wandb.ai/.../sweeps/27tb9vkv wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways. wandb: WARNING To avoid this, please fix the sweep config schema violations below: wandb: WARNING Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected) Create sweep with ID: 27m6nd76 Sweep URL: https://wandb.ai/../sweeps/27m6nd76 wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways. wandb: WARNING To avoid this, please fix the sweep config schema violations below: wandb: WARNING Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected) Create sweep with ID: b0f81ait Sweep URL: https://wandb.ai/.../sweeps/b0f81ait wandb: Agent Starting Run: sofplkis with config: wandb: learning_rate: 0.0005 wandb: lr_drop_epoch: 80 wandb: max_epoch_num: 200 wandb: optim: Adam wandb: Agent Starting Run: o5mmxwkd with config: wandb: learning_rate: 0.0003 wandb: lr_drop_epoch: 80 wandb: max_epoch_num: 100 wandb: optim: Adam wandb: Agent Starting Run: 2y8wyoo9 with config: wandb: learning_rate: 0.0001 wandb: lr_drop_epoch: 40 wandb: max_epoch_num: 150 wandb: optim: Adam wandb: Agent Starting Run: mewgkh87 with config: wandb: learning_rate: 0.0001 wandb: lr_drop_epoch: 100 wandb: max_epoch_num: 100 wandb: optim: SGD wandb: Agent Starting Run: 0ukaa50c with config: wandb: learning_rate: 0.0005 wandb: lr_drop_epoch: 100 wandb: max_epoch_num: 150 wandb: optim: SGD wandb: WARNING Malformed sweep config detected! This may cause your sweep to behave in unexpected ways. wandb: WARNING To avoid this, please fix the sweep config schema violations below: wandb: WARNING Violation 1. Additional properties are not allowed ('metrics', 'project_name' were unexpected) Create sweep with ID: lqiseifr Sweep URL: https://wandb.ai/.../sweeps/lqiseifr wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin wandb: WARNING Ignored wandb.init() arg project when running a sweep. wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin wandb: WARNING Ignored wandb.init() arg project when running a sweep. wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin wandb: WARNING Ignored wandb.init() arg project when running a sweep. wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin wandb: WARNING Ignored wandb.init() arg project when running a sweep. wandb: Agent Starting Run: mjo15l5x with config: wandb: learning_rate: 0.0004 wandb: lr_drop_epoch: 60 wandb: max_epoch_num: 200 wandb: optim: SGD wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin wandb: WARNING Ignored wandb.init() arg project when running a sweep. wandb: - Waiting for wandb.init()... wandb: - Waiting for wandb.init()... wandb: \ Waiting for wandb.init()... wandb: \ Waiting for wandb.init()... wandb: Currently logged in as: .... . Use `wandb login --relogin` to force relogin wandb: WARNING Ignored wandb.init() arg project when running a sweep. wandb: Tracking run with wandb version 0.17.1 wandb: Run `wandb offline` to turn off syncing. wandb: Syncing run 2024-06-18 02:10:19 wandb: ⭐️ View project at https://wandb.ai/... wandb: 🧹 View sweep at https://wandb.ai/.../sweeps/lqiseifr wandb: 🚀 View run at https://wandb.ai/.../runs/mjo15l5x wandb: Tracking run with wandb version 0.17.1 wandb: Run `wandb offline` to turn off syncing. wandb: Syncing run 2024-06-18 02:10:17 wandb: ⭐️ View project at https://wandb.ai/... wandb: 🧹 View sweep at https://wandb.ai/.../sweeps/31foef6f wandb: 🚀 View run at https://wandb.ai/.../runs/o5mmxwkd
> 
> From the output logs, my guess is that each GPU running the model code creates a new sweat id, and each sweat id in turn scans all the hyperparameter combinations I've set up, but I actually only need one sweat id that scans a set of hyperparameter combinations, but since I'm going to be going into my run in the sweat.agent() function part of the code function part of the code, so it's a bit of an ordering problem, but there's no solution yet!

**Comment by LTayfaker** [comment]
> Jason Davenport 评论道： 嗨，内部人员，
> 
> 由于尚未收到您的回复，我们希望就您的支持请求与您联系。如果我们可以提供进一步的帮助或您的问题已得到解决，请告知我们。
> 
> 最佳、 权重和偏差

Sorry about that, I'm busy with other things these days, I'll be the first to try out what you've suggested when it's done, and I'll be the first to let you know of any other problems or successful runs, thank you very much for your help!

**Comment by LTayfaker** [comment]
> Jason Davenport commented: Hi Internal,
> 
> We wanted to follow up with you regarding your support request as we have not heard back from you. Please let us know if we can be of further assistance or if your issue has been resolved.
> 
> Best, Weights & Biases

Sorry about that, I'm busy with other things these days, I'll be the first to try out what you've suggested when it's done, and I'll be the first to let you know of any other problems or successful runs, thank you very much for your help!
